Resource Leak Checker: calling `@CreatesMustCallFor("this")` method from constructor

Consider the following code, which appears in `checker/tests/resourceleak/LemmaStack.java`:

```java
// The Resource Leak Checker issues the following error:
// LemmaStack.java:37: error: [reset.not.owning] Calling method startProver resets the must-call
// obligations of the expression this, which is non-owning. Either annotate its declaration with an
// @Owning annotation, extract it into a local variable, or write a corresponding
// @CreatesMustCallFor annotation on the method that encloses this statement.
//     startProver();
//                ^
// 1 error

import java.io.Closeable;
import java.io.IOException;
import java.io.PrintWriter;
import org.checkerframework.checker.calledmethods.qual.EnsuresCalledMethods;
import org.checkerframework.checker.mustcall.qual.CreatesMustCallFor;
import org.checkerframework.checker.mustcall.qual.MustCall;
import org.checkerframework.checker.mustcall.qual.Owning;
import org.checkerframework.checker.nullness.qual.EnsuresNonNull;

@MustCall("close") public class LemmaStack implements Closeable {

  private @Owning @MustCall("close") PrintWriter session;

  @CreatesMustCallFor("this")
  @EnsuresNonNull("session")
  private void startProver() {
    try {
      if (session != null) {
        session.close();
      }
      session = new PrintWriter("filename.txt");
    } catch (IOException e) {
      throw new Error(e);
    }
  }

  public LemmaStack() {
    startProver();
  }

  @EnsuresCalledMethods(value = "session", methods = "close")
  @Override
  public void close(LemmaStack this) {
    session.close();
  }
}
```

None of the three suggestions in the error messages works.
 * annotate its declaration with an `@Owning` annotation:  there is no declaration of `this` to annotate
 * extract it into a local variable :  `@Owning` cannot be written on a local variable
 * write a corresponding `@CreatesMustCallFor` annotation on the method that encloses this statement:  that does not work for the constructor.

The message should be improved to only make suggestions that are possible.
Furthermore, this false negative warning should be eliminated.

--- a/checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java
+++ b/checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java
@@ -562,8 +562,12 @@ class MustCallConsistencyAnalyzer {
    * Checks that an invocation of a CreatesMustCallFor method is valid.
    *
    * <p>Such an invocation is valid if any of the conditions in {@link
-   * #isValidCreatesMustCallForExpression(Set, JavaExpression, TreePath)} is true. If none of these
-   * conditions are true, this method issues a reset.not.owning error.
+   * #isValidCreatesMustCallForExpression(Set, JavaExpression, TreePath)} is true for each
+   * expression in the argument to the CreatesMustCallFor annotation. As a special case, the
+   * invocation of a CreatesMustCallFor method with "this" as its expression is permitted in the
+   * constructor of the relevant class (invoking a constructor already creates an obligation). If
+   * none of these conditions are true for any of the expressions, this method issues a
+   * reset.not.owning error.
    *
    * <p>For soundness, this method also guarantees that if any of the expressions in the
    * CreatesMustCallFor annotation has a tracked Obligation, any tracked resource aliases of it will
@@ -593,6 +597,14 @@ class MustCallConsistencyAnalyzer {
     if (missing.isEmpty()) {
       // All expressions matched one of the rules, so the invocation is valid.
       return;
+    }
+
+    // Special case for invocations of CreatesMustCallFor("this") methods in the constructor.
+    if (missing.size() == 1) {
+      JavaExpression expression = missing.get(0);
+      if (expression instanceof ThisReference && TreePathUtil.inConstructor(currentPath)) {
+        return;
+      }
     }
 
     String missingStrs = StringsPlume.join(", ", missing);

--- a/checker/tests/resourceleak/LemmaStack.java
+++ b/checker/tests/resourceleak/LemmaStack.java
@@ -1,5 +1,4 @@
 // Test case for issue #5175: https://github.com/typetools/checker-framework/issues/5175
-// @skip-test until the issue is fixed
 
 // The Resource Leak Checker issues the following error:
 // LemmaStack.java:40: error: [reset.not.owning] Calling method startProver resets the must-call

